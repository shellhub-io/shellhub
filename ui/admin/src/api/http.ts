import { AxiosInstance } from "axios";
import { setupInterceptorsTo } from "./interceptors";
import * as axiosTs from "./client";
import { Configuration } from "./client";
import { BaseAPI } from "./client/base";

const configuration = new Configuration();
configuration.basePath = `${window.location.protocol}//${window.location.host}`;
configuration.accessToken = localStorage.getItem("cloud_token") || "";

// eslint-disable-next-line import/no-mutable-exports
let adminApi = new axiosTs.AdminApi(configuration);
// eslint-disable-next-line import/no-mutable-exports
let cloudApi = new axiosTs.CloudApi(configuration);

export const createNewClient = () => {
  const newConfiguration = new Configuration();
  newConfiguration.basePath = `${window.location.protocol}//${window.location.host}`;
  newConfiguration.accessToken = localStorage.getItem("cloud_token") || "";
  adminApi = new axiosTs.AdminApi(newConfiguration);
  cloudApi = new axiosTs.CloudApi(newConfiguration);
  return { adminApi, cloudApi };
};

declare module "./client/base" {
  interface BaseAPI {
    getAxios(): AxiosInstance;
    getConfiguration(): Configuration | undefined;
    setConfiguration(configuration: Configuration): void;
  }
}

/** Returns the axios instance */
BaseAPI.prototype.getAxios = function getAxios(this: BaseAPI): AxiosInstance {
  return this.axios;
};

/** Returns the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.getConfiguration = function getConfiguration(this: BaseAPI): Configuration | undefined {
  return this.configuration;
};

/** Sets the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.setConfiguration = function setConfiguration(this: BaseAPI, configuration: Configuration): void {
  this.configuration = configuration;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
setupInterceptorsTo(adminApi.axios);

export { adminApi, cloudApi };
