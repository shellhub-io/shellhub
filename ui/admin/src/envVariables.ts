/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
