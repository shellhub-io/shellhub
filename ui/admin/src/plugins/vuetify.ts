// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
