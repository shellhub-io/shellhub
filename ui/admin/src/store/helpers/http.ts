import Axios from "axios";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore //* use while route mapping is not in TypeScript
export default () => {
  const axios = Axios.create({
    baseURL: `${window.location.protocol}//${window.location.host}/admin`,
    headers: {
      Authorization: `Bearer ${localStorage.getItem("cloud_token")}`,
    },
  });

  return axios;
};
