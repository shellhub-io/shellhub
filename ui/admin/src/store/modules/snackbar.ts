import { Module } from "vuex";
import { State } from "./../index";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,

  state: {
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: { typeMessage: "", typeContent: "" },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarNoContent: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarSuccessDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorCustom: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarError = true;
    },

    setSnackbarErrorLicense: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = { typeMessage: "", typeContent: value };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarSuccessAction", data);
    },

    showSnackbarSuccessDefault: ({ commit }) => {
      commit("setSnackbarSuccessDefault");
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = { typeMessage: "loading", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarNoContent: ({ commit }) => {
      const data = { typeMessage: "no-content", typeContent: "" };
      commit("setSnackbarNoContent", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorCustom: ({ commit }, value) => {
      const data = { typeMessage: "custom", typeContent: value };
      commit("setSnackbarErrorCustom", data);
    },

    showSnackbarErrorLicense: ({ commit }, value) => {
      const data = { typeMessage: "licenseRequired", typeContent: value };
      commit("setSnackbarErrorLicense", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
