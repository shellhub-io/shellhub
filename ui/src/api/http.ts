import axios, { AxiosInstance } from "axios";
import * as axiosTs from "./client";
import { Configuration } from "./client";
import { BaseAPI } from "./client/base";
import { setupInterceptorsTo } from "./interceptors";

// This is the default configuration for local instance endpoints
const configuration = new Configuration();
configuration.basePath = `${window.location.protocol}//${window.location.host}`;
configuration.accessToken = localStorage.getItem("token") || "";

// We need a custom configuration for cloud endpoints
const cloudApiConfiguration = new Configuration();
cloudApiConfiguration.basePath = "https://cloud.shellhub.io";

// Creates a new axios instance and setup interceptors by default
const newAxiosInstance = (setupInterceptor = true): AxiosInstance => {
  const instance = axios.create();
  if (setupInterceptor) setupInterceptorsTo(instance);
  return instance;
};
const sessionsApi = new axiosTs.SessionsApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const devicesApi = new axiosTs.DevicesApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const containersApi = new axiosTs.ContainersApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const systemApi = new axiosTs.SystemApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const namespacesApi = new axiosTs.NamespacesApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const apiKeysApi = new axiosTs.ApiKeysApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const sshApi = new axiosTs.SshApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const tagsApi = new axiosTs.TagsApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const usersApi = new axiosTs.UsersApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const mfaApi = new axiosTs.MfaApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const tunnelApi = new axiosTs.TunnelsApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const billingApi = new axiosTs.BillingApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const rulesApi = new axiosTs.RulesApi(
  configuration,
  undefined,
  newAxiosInstance(),
);
const announcementApi = new axiosTs.AnnouncementsApi(
  cloudApiConfiguration,
  undefined,
  newAxiosInstance(false),
);

/**
 * Extends the interface BaseAPI to include a new method `getAxios` allowing
 * access to the protected axios property without the need to access it
 * directly from outside the class and avoiding a linting error caused by
 * accessing a protected property.
 **/
declare module "./client/base" {
  interface BaseAPI {
    getAxios(): AxiosInstance;
    getConfiguration(): Configuration | undefined;
    setConfiguration(configuration: Configuration): void;
  }
}

/** Returns the axios instance */
BaseAPI.prototype.getAxios = function getAxios(this: BaseAPI): AxiosInstance {
  return this.axios;
};

/** Returns the configuration */
BaseAPI.prototype.getConfiguration = function getConfiguration(this: BaseAPI): Configuration | undefined {
  return this.configuration;
};

/** Sets the configuration */
BaseAPI.prototype.setConfiguration = function setConfiguration(this: BaseAPI, configuration: Configuration): void {
  this.configuration = configuration;
};

// Reloads the configuration for all APIs
const reloadConfiguration = () => {
  [
    sessionsApi,
    devicesApi,
    containersApi,
    systemApi,
    namespacesApi,
    apiKeysApi,
    sshApi,
    tagsApi,
    usersApi,
    mfaApi,
    tunnelApi,
    billingApi,
    rulesApi,
  ].forEach((api) => {
    api.setConfiguration(configuration);
  });
};

export {
  configuration,
  reloadConfiguration,
  sessionsApi,
  devicesApi,
  containersApi,
  systemApi,
  namespacesApi,
  apiKeysApi,
  sshApi,
  tagsApi,
  usersApi,
  mfaApi,
  tunnelApi,
  billingApi,
  rulesApi,
  announcementApi,
};
