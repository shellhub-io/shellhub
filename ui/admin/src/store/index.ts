import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
