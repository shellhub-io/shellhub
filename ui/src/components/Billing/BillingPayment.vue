<template>
  <v-row>
    <v-col>
      <v-text-field
        v-model="customer.name"
        label="Name"
        disabled
        data-test="customer-name" />
    </v-col>
    <v-col>
      <v-text-field
        v-model="customer.email"
        label="E-mail"
        disabled
        data-test="customer-email" />
    </v-col>
  </v-row>
  <v-col class="ma-0 pa-0">
    <h4 data-test="credit-card-text">Your credit cards</h4>
  </v-col>
  <v-list
    v-if="!(customer.payment_methods == null)"
    data-test="payment-methods-list"
    nav
    bg-color="transparent"
    class="w-100 pa-0 pt-2 content-card">
    <v-col v-for="(item, i) in customer.payment_methods" :key="i" class="pa-0 mt-2 mb-2">
      <v-card :key="i" class="bg-v-theme-card" variant="flat">
        <v-list-item :key="i" data-test="payment-methods-item" @click="setDefaultPayment(item.id)" class="card">
          <v-row align="center" cols="12">
            <v-col cols="1">
              <BillingIcon :icon-name="item.brand" />
            </v-col>
            <v-col cols="4">
              <b>{{ item.number }}</b>
            </v-col>
            <v-col cols="2">
              {{ item.exp_month + "/" + item.exp_year }}
            </v-col>
            <v-col cols="2">
              {{ item.cvc }}
            </v-col>
            <v-col cols="3" v-if="item.default === true">
              <v-chip density="comfortable">
                <b>default</b>
                <v-tooltip
                  activator="parent"
                  location="top"
                >This payment method will be used on your subscription</v-tooltip>
              </v-chip>
            </v-col>
            <v-col cols="3" v-if="!item.default" class="d-flex flex-column align-end">
              <v-btn variant="text" icon="mdi-delete" data-test="payment-methods-delete-btn" @click.stop="deletePaymentMethod(item.id)" />
            </v-col>
          </v-row>
        </v-list-item>
      </v-card>
    </v-col>
  </v-list>
  <v-col v-else>
    <v-card-subtitle>You don't have any registered cards yet, please add one</v-card-subtitle>
  </v-col>
  <v-row class="mt-0 pt-0 pt-0" cols="12">
    <v-col cols="8" class="pr-0">
      <StripeElements
        v-if="stripeLoaded && addNewCard"
        v-slot="{ elements }"
        ref="elms"
        :stripe-key="stripeKey"
        :instance-options="instanceOptions"
        :elements-options="elementsOptions">
        <StripeElement
          type="card"
          ref="card"
          :elements="elements"
          :options="cardOptions" />
      </StripeElements>
    </v-col>
    <v-col cols="4" class="d-flex flex-lg-column align-center justify-center pt-0">
      <v-btn
        @click='addNewCard ? savePayment() : addNewCard = true'
        :text='addNewCard ? "Save card" : "Add new card "'
        prepend-icon="mdi-credit-card-plus"
        color="primary"
        data-test="add-card-btn" />
    </v-col>
  </v-row>
  <v-row>
    <v-col>
      <v-alert v-if="alertRender" icon="$error" :text="errorMessage" type="error" data-test="alert-message" />
    </v-col>
  </v-row>
</template>

<script lang='ts' setup>
import { onBeforeMount, onMounted, ref, reactive, computed } from "vue";
import { loadStripe } from "@stripe/stripe-js";
import { StripeElements, StripeElement } from "vue-stripe-js";
import type { StripeConstructorOptions, StripeElementsOptions, StripeCardElementOptions } from "@stripe/stripe-js";
import BillingIcon from "./BillingIcon.vue";
import { useStore } from "@/store";
import handleError from "@/utils/handleError";
import { envVariables } from "@/envVariables";
import { Customer } from "@/interfaces/ICustomer";

const emit = defineEmits(["no-payment-methods", "has-default-payment", "customer-id-created"]);
const stripeKey = computed(() => envVariables.stripeKey);
const stripeLoaded = ref(false);
const customer: Customer = reactive({ name: "", email: "", payment_methods: [] });
const store = useStore();
const consumerData = computed(() => store.getters["customer/getCustomer"]);
const card = ref();
const elms = ref();
const errorMessage = ref();
const alertRender = ref(false);
const addNewCard = ref(false);
const namespace = computed(() => store.getters["namespaces/get"]);
const instanceOptions = ref<StripeConstructorOptions>({
});

const elementsOptions = ref<StripeElementsOptions>({
  fonts: [{ cssSrc: "https://fonts.googleapis.com/css2?family=Lato" }],
  appearance: {
    theme: "night",
  },
  locale: "en",
  loader: "auto",
  mode: "setup",
  currency: "usd",
  payment_method_creation: "manual",
  // https://stripe.com/docs/js/elements_object/create#stripe_elements-options
});
const cardOptions = ref<StripeCardElementOptions>({
  style: {
    base: {
      iconColor: "#c4f0ff",
      color: "#fff",
      backgroundColor: "#22252B",
      lineHeight: "50px",
      padding: "12px",
      fontWeight: "400",
      fontFamily: "Roboto, Open Sans, Segoe UI, sans-serif",
      fontSize: "16px",
      fontSmoothing: "antialiased",
      ":-webkit-autofill": {
        color: "white",
      },
      "::placeholder": {
        color: "white",
      },
    },
    invalid: {
      iconColor: "#FFC7EE",
      color: "#FFC7EE",
    },
  },
  hidePostalCode: true,
});

const fetchData = async () => {
  try {
    await store.dispatch("customer/fetchCustomer");
    const customerDetails = consumerData.value.data;

    Object.assign(customer, {
      name: customerDetails?.name || "",
      email: customerDetails?.email || "",
      payment_methods: customerDetails?.payment_methods,
    });

    if (customerDetails.payment_methods?.length === 0 || customerDetails.payment_methods === null) {
      emit("no-payment-methods");
    } else {
      emit("has-default-payment");
    }
  } catch (error) {
    handleError(error);
  }
};

class CustomError {
  code: string;

  message: string;

  constructor(code: string, message: string) {
    this.code = code;
    this.message = message;
  }
}

const savePayment = async () => {
  const cardElement = card.value.stripeElement;
  elms.value.instance.createPaymentMethod({
    type: "card",
    card: cardElement,
  })
    .then(async (result: { paymentMethod: { id: string; }; }) => {
      try {
        const id: string = result.paymentMethod.id || "";
        await store.dispatch("customer/attachPaymentMethod", id);
        await fetchData();
        alertRender.value = false;
        addNewCard.value = false;
      } catch (error) {
        const errorMessages = {
          card_declined: "Your payment method was declined, check if your card is valid or has sufficient funds",
          expired_card: "Your payment was declined because the card has expired.",
          incorrect_cvc: "Your payment was declined due to an incorrect CVC.",
          processing_error: "Your payment was declined due to a processing error.",
          incorrect_number: "Your payment was declined because the card number is incorrect.",
          default: "An error occurred during payment processing.",
        };

        alertRender.value = true;
        const isMessageError = (error: unknown): error is CustomError => typeof error === "object" && error !== null && "code" in error;
        errorMessage.value = isMessageError(error) ? errorMessages[error.code] : errorMessages.default;
      }
    });
};

const setDefaultPayment = async (id: string) => {
  try {
    await store.dispatch("customer/setDefaultPaymentMethod", id).then(async () => {
      await fetchData();
    });
  } catch (error) {
    handleError(error);
  }
};

const deletePaymentMethod = async (id: string) => {
  try {
    await store.dispatch("customer/detachPaymentMethod", id).then(async () => {
      await fetchData();
    });
  } catch (error) {
    handleError(error);
  }
};

onMounted(async () => {
  const tenant = computed(() => localStorage.getItem("tenant"));
  await store.dispatch("namespaces/get", tenant.value);
  if (namespace.value.billing == null || namespace.value.billing.customer_id === "") {
    try {
      await store.dispatch("customer/createCustomer");
      emit("customer-id-created");
    } catch (error) {
      handleError(error);
    }
  }
  await fetchData();
});

onBeforeMount(async () => {
  const stripePromise = loadStripe(stripeKey.value);
  await stripePromise.then(() => {
    stripeLoaded.value = true;
  });
});
</script>

<style scoped>
.card:hover,
.card:focus {
  border-left: 5px solid #7284d0;
  border-right: 5px solid #7284d0;
  transition: ease-in-out 200ms;
}
.content-card {
  max-height: 45vh;
  overflow: auto;
}

::-webkit-scrollbar {
    width: 6px;
  }
    ::-webkit-scrollbar-track {
      background-color: rgb(255 255 255 / 10%);
      border-radius: 10px;
    }
    ::-webkit-scrollbar-thumb {
      background-color: rgb(0 0 0 / 80%);
      border-radius: 10px;
    }
</style>
