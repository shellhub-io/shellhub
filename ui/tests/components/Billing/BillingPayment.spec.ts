import { flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { createVuetify } from "vuetify";
import MockAdapter from "axios-mock-adapter";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { billingApi, namespacesApi } from "@/api/http";
import { store, key } from "@/store";
import BillingPayment from "@/components/Billing/BillingPayment.vue";
import { envVariables } from "@/envVariables";
import { SnackbarPlugin } from "@/plugins/snackbar";

describe("Billing Payment", () => {
  let wrapper: VueWrapper<InstanceType<typeof BillingPayment>>;

  const vuetify = createVuetify();

  let mockNamespace: MockAdapter;

  let mockCustomer: MockAdapter;

  const members = [
    {
      id: "xxxxxxxx",
      type: "owner",
      username: "test",
    },
    {
      id: "xxxxxxxy",
      type: "observer",
      username: "test2",
    },
  ];

  const billingData = {
    active: false,
    status: "inactive",
    customer_id: "cus_test123",
    subscription_id: "sub_test",
    current_period_end: 123781839,
    created_at: "",
    updated_at: "",
  };

  const namespaceData = {
    name: "test",
    owner: "test",
    tenant_id: "fake-tenant-data",
    members,
    max_devices: 3,
    devices_count: 3,
    created_at: "",
    billing: billingData,
  };

  const customerData = {
    id: "cus_test123",
    name: "test",
    email: "test@test.com",
    payment_methods: [
      {
        id: "pm_test123",
        number: "1234 1234 1234 1234",
        brand: "visa",
        exp_month: 12,
        exp_year: 2999,
        cvc: "123",
        default: false,
      },
      {
        id: "pm_test456",
        number: "1234 1234 1234 5678",
        brand: "mastercard",
        exp_month: 12,
        exp_year: 2999,
        cvc: "123",
        default: true,
      },
    ],
  };

  beforeEach(() => {
    vi.useFakeTimers();
    localStorage.setItem("tenant", "fake-tenant-data");
    envVariables.isCloud = true;

    mockCustomer = new MockAdapter(billingApi.getAxios());
    mockNamespace = new MockAdapter(namespacesApi.getAxios());

    mockNamespace.onGet("http://localhost:3000/api/namespaces/fake-tenant-data").reply(200, namespaceData);
    mockCustomer.onGet("http://localhost:3000/api/billing/customer").reply(200, customerData);

    store.commit("namespaces/setNamespace", namespaceData);

    wrapper = mount(BillingPayment, {
      global: {
        plugins: [[store, key], vuetify, SnackbarPlugin],
      },
      mocks: {
        $stripe: {
          elements: () => ({
            create: () => ({
              mount: () => null,
            }),
          }),
        },
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Renders the correct html", async () => {
    await flushPromises();
    expect(wrapper.findComponent('[data-test="customer-name"]').exists()).toBe(true);
    expect(wrapper.findComponent('[data-test="customer-email"]').exists()).toBe(true);
    expect(wrapper.findComponent('[data-test="payment-methods-list"]').exists()).toBe(true);
    expect(wrapper.findComponent('[data-test="payment-methods-item"]').exists()).toBe(true);
    expect(wrapper.findComponent('[data-test="payment-methods-delete-btn"]').exists()).toBe(true);
    expect(wrapper.findComponent('[data-test="add-card-btn"]').exists()).toBe(true);
    expect(wrapper.findComponent('[data-test="alert-message"]').exists()).toBe(false);
  });

  it("Detach payment method", async () => {
    await flushPromises();
    mockCustomer.onPost("http://localhost:3000/api/billing/paymentmethod/detach").reply(200);

    const detachPaymentMethodSpy = vi.spyOn(store, "dispatch");

    await wrapper.findComponent('[data-test="payment-methods-delete-btn"]').trigger("click");

    expect(detachPaymentMethodSpy).toHaveBeenCalledWith("customer/detachPaymentMethod", "pm_test123");
  });

  it("Set default payment method", async () => {
    await flushPromises();
    mockCustomer.onPost("http://localhost:3000/api/billing/paymentmethod/default").reply(200);

    const defaultPaymentMethodSpy = vi.spyOn(store, "dispatch");

    await wrapper.findComponent('[data-test="payment-methods-item"]').trigger("click");

    expect(defaultPaymentMethodSpy).toHaveBeenCalledWith("customer/setDefaultPaymentMethod", "pm_test123");
  });

  // TODO STRIPE TEST SAVE PAYMENT METHOD
  // it("Saves the payment method", async () => {
  //   mockCustomer.onPost("http://localhost:3000/api/billing/paymentmethod/attach", { stripeTestCard }).reply(200);
  //   const addPaymentMethodSpy = vi.spyOn(store, "dispatch");
  //   wrapper.vm.addNewCard = true;
  //   await wrapper.find('[data-test="add-card-btn"]').trigger("click");
  //   vi.runOnlyPendingTimers();
  //   expect(addPaymentMethodSpy).toHaveBeenCalledWith("customer/attachPaymentMethod", stripeTestCard);
  // });

  // it("Fails to save the payment method", async () => {
  //   await flushPromises();
  //   mockCustomer.onPost("http://localhost:3000/api/billing/paymentmethod/attach").reply(424);
  //   wrapper.vm.addNewCard = true;
  //   const addPaymentMethodSpy = vi.spyOn(store, "dispatch");
  //   await wrapper.findComponent('[data-test="payment-methods-item"]').trigger("click");
  // });
});
