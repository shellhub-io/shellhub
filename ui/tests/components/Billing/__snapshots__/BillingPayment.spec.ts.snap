// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Billing Payment > Renders the component 1`] = `
"<div data-v-ec12e293="" class="v-row">
  <div data-v-ec12e293="" class="v-col">
    <div data-v-ec12e293="" class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--disabled v-text-field" data-test="customer-name">
      <!---->
      <div class="v-input__control">
        <div class="v-field v-field--center-affix v-field--disabled v-field--variant-filled v-theme--light v-locale--is-ltr">
          <div class="v-field__overlay"></div>
          <div class="v-field__loader">
            <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
              <!---->
              <div class="v-progress-linear__background"></div>
              <div class="v-progress-linear__buffer" style="width: 0%;"></div>
              <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                <div class="v-progress-linear__indeterminate">
                  <div class="v-progress-linear__indeterminate long"></div>
                  <div class="v-progress-linear__indeterminate short"></div>
                </div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
          <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-v-0">
              <!---->Name
            </label><label class="v-label v-field-label" for="input-v-0">
              <!---->Name
            </label>
            <!----><input disabled="" size="1" type="text" id="input-v-0" aria-describedby="input-v-0-messages" class="v-field__input" value="">
            <!---->
          </div>
          <!---->
          <!---->
          <div class="v-field__outline">
            <!---->
            <!---->
          </div>
        </div>
      </div>
      <!---->
      <div id="input-v-0-messages" class="v-input__details" role="alert" aria-live="polite">
        <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
          <!---->
        </transition-group-stub>
        <!---->
      </div>
    </div>
  </div>
  <div data-v-ec12e293="" class="v-col">
    <div data-v-ec12e293="" class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--disabled v-text-field" data-test="customer-email">
      <!---->
      <div class="v-input__control">
        <div class="v-field v-field--center-affix v-field--disabled v-field--variant-filled v-theme--light v-locale--is-ltr">
          <div class="v-field__overlay"></div>
          <div class="v-field__loader">
            <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
              <!---->
              <div class="v-progress-linear__background"></div>
              <div class="v-progress-linear__buffer" style="width: 0%;"></div>
              <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                <div class="v-progress-linear__indeterminate">
                  <div class="v-progress-linear__indeterminate long"></div>
                  <div class="v-progress-linear__indeterminate short"></div>
                </div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
          <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-v-2">
              <!---->E-mail
            </label><label class="v-label v-field-label" for="input-v-2">
              <!---->E-mail
            </label>
            <!----><input disabled="" size="1" type="text" id="input-v-2" aria-describedby="input-v-2-messages" class="v-field__input" value="">
            <!---->
          </div>
          <!---->
          <!---->
          <div class="v-field__outline">
            <!---->
            <!---->
          </div>
        </div>
      </div>
      <!---->
      <div id="input-v-2-messages" class="v-input__details" role="alert" aria-live="polite">
        <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
          <!---->
        </transition-group-stub>
        <!---->
      </div>
    </div>
  </div>
</div>
<div data-v-ec12e293="" class="v-col ma-0 pa-0">
  <h4 data-v-ec12e293="" data-test="credit-card-text">Your credit cards</h4>
</div>
<div data-v-ec12e293="" class="v-list v-list--nav v-theme--light bg-transparent v-list--density-default v-list--one-line w-100 pa-0 pt-2 content-card" tabindex="0" role="listbox" data-test="payment-methods-list"></div>
<div data-v-ec12e293="" class="v-row mt-0 pt-0 pt-0" cols="12">
  <div data-v-ec12e293="" class="v-col v-col-8 pr-0">
    <!--v-if-->
  </div>
  <div data-v-ec12e293="" class="v-col v-col-4 d-flex flex-lg-column align-center justify-center pt-0"><button data-v-ec12e293="" type="button" class="v-btn v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated" data-test="add-card-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="mdi-credit-card-plus mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span><span class="v-btn__content" data-no-activator="">Add new card </span>
      <!---->
      <!---->
    </button></div>
</div>
<div data-v-ec12e293="" class="v-row">
  <div data-v-ec12e293="" class="v-col">
    <!--v-if-->
  </div>
</div>"
`;
